framework/src/auxkernels/NodalPatchRecovery.C

```diff
@@ -0,0 +1,208 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "NodalPatchRecovery.h"
+#include "SwapBackSentinel.h"
+#include "libmesh/utility.h"
+
+template <>
+InputParameters
+validParams<NodalPatchRecovery>()
+{
+  InputParameters params = validParams<AuxKernel>();
+  params.addRequiredParam<unsigned>("order", "polynomial order");
+  params.set<bool>("use_nodal_patch_recovery") = true;
+  return params;
+}
+
+NodalPatchRecovery::NodalPatchRecovery(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _order(getParam<unsigned>("order")),
+    _dim(_mesh.dimension()),
+    _fe_problem(*parameters.get<FEProblemBase *>("_fe_problem_base")),
+    _multi_index(multiIndex(_dim, _order))
+{
+  MeshBase & meshhelper = _mesh.getMesh();
+  meshhelper.allow_renumbering(false);
+  for (const auto & elem :
+       as_range(meshhelper.semilocal_elements_begin(), meshhelper.semilocal_elements_end()))
```

Comment:
I take it that this fixed the parallel problems? Are you essentially ghosting out a second layer of elements?

Reply:
You are right I am fixing the parallel issue here. However, this is not ghosting out a second layer.
It is a bit complicated to explain without drawing any graph. Here I am ghosting out all semilocal elements -- which is different from simply ghosting one layer out in a very subtle way.

I am also trying to get distributed mesh to work using RelationshipManager.

Reply:
I recommend you stay away from semilocal elements. It's not quite what you think it is, trust me. Try searching for it in the tickets! 

The right way is with a relationship manager as you've already said. There's already existing objects that will do arbitrary numbers of layers or point neighbors and then you are good to go.

Reply:
@hugary1995 are you trying to get _point neighbors_ here? Because that would require at least *2* additional ghosting layers with quad elements, and possibly more with tets etc! You also need to be super careful with periodic boundary conditions.

Reply:
This problem is solved: See how I do it for the GrainTracker:
https://github.com/idaholab/moose/blob/devel/modules/phase_field/src/postprocessors/GrainTracker.C#L55

https://github.com/idaholab/moose/blob/devel/modules/phase_field/src/relationshipmanagers/GrainTrackerHaloRM.C

Now, your use case will be a little more complex because you actually need solution data off from the ghosted element correct? For that, you'll need to create an AlgebraicRelationshipManager and we'll have to figure out exactly what we'll need to do to get solution information. This is probably also a solved problem but if not, I'm willing to help you reach that solution. Please stop by and chat when you are ready to work on this. @roystgnr will be helpful in well in telling us 1) if a ghosting functor exists that already gets solution information from n layers deep 2) Is that built into the "DefaultCoupling" ghosting functor already? Sorry, too lazy to look at the moment.

Reply:
It depends on how you define "layers".  DefaultCoupling and PointNeighborCoupling both have set_n_levels(int) methods; if you set that to e.g. 2 then with DefaultCoupling you get side-neighbors-of-side-neighbors and with PointNeighborCoupling you get point-neighbors-of-point-neighbors.

Reply:
@permcody I'll have time to work on this next week. What would be the best time for you to meet?